File: Core/src/services/domain_service.py
import grpc

from configs.redis_config import RedisConfig
from controllers.domain_controller import DomainController
from proto.objects_pb2 import Domain, GetDomainsPoolRequest, GetDomainsPoolResponse
from proto.objects_pb2_grpc import DomainServiceServicer


class DomainService(DomainServiceServicer):
    async def GetDomainsPool(
            self,
            request: GetDomainsPoolRequest,
            context: grpc.ServicerContext) -> GetDomainsPoolResponse:
        async with RedisConfig() as session:
            domains = [
                Domain(name=domain_name)
                for domain_name in await DomainController(session).get_all_domains()
            ]
        return GetDomainsPoolResponse(domains=domains)

File: core/internal/grpc/domain_service/domain_service.go
package domain_service

import (
	"context"

	proto "github.com/mekstack/nataas/core/proto/pb"

	"google.golang.org/grpc"
)

type DomainController interface {
	GetDomainsPool(context.Context) ([]*proto.Domain, error)
}

type service struct {
	proto.UnimplementedDomainServiceServer
	domainController DomainController
}

func Register(server *grpc.Server, controller DomainController) {
	proto.RegisterDomainServiceServer(server, &service{
		domainController: controller,
	})
}

func (d *service) GetDomainsPool(ctx context.Context, request *proto.GetDomainsPoolRequest) (*proto.GetDomainsPoolResponse, error) {
	pool, err := d.domainController.GetDomainsPool(ctx)

	if err != nil {
		return nil, err
	}

	return &proto.GetDomainsPoolResponse{
		Domains: pool,
	}, nil
}
